app/models/labware_creators/stamped_plate_adding_randomised_controls.rb

```diff
@@ -0,0 +1,255 @@
+    end
+
+    # used to fetch the sample cohort from a parent well, for use in writing to the control sample metadata
+    # we are assuming this contains a generic value shared by all samples in the parent plate
+    def control_cohort
+      @control_cohort ||= parent_wells_with_aliquots.first.aliquots.first.sample.sample_metadata.cohort
+    end
+
+    # fetch the api v2 study object for the control study name from the purpose config
+    def control_study_v2
+      @control_study_v2 ||= Sequencescape::Api::V2::Study.find(name: control_study_name).first
+    end
+
+    # fetch the api v2 project object for the control project name from the purpose config
+    def control_project_v2
+      @control_project_v2 ||= Sequencescape::Api::V2::Project.find(name: control_project_name).first
+    end
+
+    # this transfer collection stamps all the samples from the parent into the child plate,
+    # except for those being displaced by controls (uses labware_wells method via well_filter)
+    def transfer_material_from_parent!
+      api.transfer_request_collection.create!(user: user_uuid, transfer_requests: transfer_request_attributes)
+    end
+
+    def transfer_request_attributes
+      well_filter.filtered.map do |well, additional_parameters|
+        request_hash(well, @child_plate_v2, additional_parameters)
+      end
+    end
+
+    # create the control sample, setting the sample name and metadata, then create
+    # an aliquot containing the control and link it to the selected child well
+    def create_control_in_child_well(control, child_well_v2, well_location)
+      # check the well should be empty
+      unless child_well_v2.aliquots.length.zero?
+        errors.add(:base, "Expecting child plate well to be empty at location #{well_location}")
+      end
+
+      control_v2 = create_control_sample(control, well_location)
+
+      update_control_sample_metadata(control_v2, well_location)
+
+      create_aliquot_in_child_well(control_v2, child_well_v2, well_location)
+    end
+
+    # create the control sample and metadata NB. sample_name cannot contain spaces!!
+    def create_control_sample(control, well_location)
+      sample_name = "#{control.name_prefix}#{control_desc}_#{well_location}"
+      control_v2 =
+        Sequencescape::Api::V2::Sample.new(
+          name: sample_name,
+          sanger_sample_id: sample_name,
+          control: true,
+          control_type: control.control_type
+        )
+      control_v2.relationships.studies = [control_study_v2]
+
+      return control_v2 if control_v2.save
+
+      errors.add(:base, "New control (type #{control.control_type}) did not save for location #{well_location}")
+    end
+
+    def update_control_sample_metadata(control_v2, well_location)
+      if control_v2.sample_metadata.update(
+           supplier_name: control_v2.name,
+           cohort: control_cohort,
+           sample_description: control_desc
```

Comment:
all controls will have the same sample description? if so, why do we give a value? is it needed?

Reply:
sample description is coming from one of the other samples, its the specimen plate barcode for Bioscan (the source insect plate barcode coming in from the manifest) so it is the same in all wells and needed for the mBrave file creation